api 서버 구현

rave189 opened the pull request.

qwlake commented on knu-notice-server/src/main/kotlin/knu/notice/knunoticeserver/domain/Support.kt:
```diff
@@ -0,0 +1,16 @@
+@Table(name = "support")
+data class Support(
+    @Id
+    val id: Int,
```
int로 하신 특별한 이유가 있나요?

qwlake commented on knu-notice-server/src/main/kotlin/knu/notice/knunoticeserver/service/PushAlarmService.kt:
```diff
@@ -0,0 +1,127 @@
+        }
+
+        // 가장 마지막 공지를 PushStatus에 다시 저장한다.
+        pushStatusRepository.save(PushStatus(1, lastNoticeId, LocalDateTime.now()))
```
updated_at 같은 필드의 경우 코드에서 직접 넣지 않아도 `jpa auditing`으로 자동으로 들어가도록 할 수 있습니다. 관련 키워드를 검색해보세요.
이 부분은 굳이 본 pr에서 수정되지 않아도 문제 없지만, 추후에는 꼭 수정하시면 좋을 것 같습니다.

qwlake replied:
id를 1로 세팅하고 저장하는데, 이렇게 하면 기존 데이터가 덮어씌워지나요? 아니면 저장에 실패하나요?
그리고 매번 저장할 때마다 id가 1로 고정되는건가요?

qwlake commented:
pagination 할 때 db에서 모두 가져온 후, 코드상에서 자르는 것 같아 보입니다.
그런 것 보다는 쿼리를 날릴 때부터 pagination 하는 것을 권장드립니다.
https://javabydeveloper.com/spring-jdbctemplate-pagination-examples/

이 부분은 굳이 본 pr에서 수정되지 않아도 문제 없지만, 추후에는 꼭 수정하시면 좋을 것 같습니다.

rave189 commented:
> pagination 할 때 db에서 모두 가져온 후, 코드상에서 자르는 것 같아 보입니다. 그런 것 보다는 쿼리를 날릴 때부터 pagination 하는 것을 권장드립니다. https://javabydeveloper.com/spring-jdbctemplate-pagination-examples/

rave189 committed:
add project


1. Support 기능 추가, Domain과 DTO 분리, Pagination 수정, 코드 정리
추가
- Support Repository, Domain, DTO 클래스 추가

수정
- Domain과 DTO를 분리
- Pagination이 2번째 페이지부터 시작하던 현상 수정

2. PushAlarm 컨트롤러, 서비스 파일 및 firebase 의존성 추가
추가
- PUshAlarm 컨트롤러, 서비스 파일
- firebase 의존성

수정
- resources아래 파일을 모두 무시하도록 gitignore 수정

3. 공지사항 추가
추가
- 학과 정보를 받아 해당 학과에 대한 공지사항을 받아오는 메소드 추가
- 학과 정보와 키워드 정보를 받아 해당 학과에 키워드가 하나라도 들어간 공지 사항을 받아노는 메소드 추가
- JpaRepository로는 위의 메소드를 구현할 수 없어 JdbcTemplate을 사용하는 Repository 클래스 추가
- Util 파일 추가로 count, next, prev를 하나의 메소드에서 생성하고 BaseResponse를 반환하는 메소드 추가
- fcm 관련 라이브러리 gradle에 추가

4. push alarm기능 추가 및 ExceptionHandler 추가
추가
- Fcm 관련 Configuration 클래스 추가
- 이전까지 보낸 공지사항 정보를 저장하는 PushStatus 추가 및 Repository 
추가
- 새로운 공지사항이 들어오면 push alarm 보내는 기능 추가
- Exception catch 코드 및 handle 코드 추가
- Custom Exception 클래스 추가

수정
- 전체 Device 또는 User 데이터를 페이지 단위로 반환하는 클래스를 Id를 받아 해당하는 Device 또는 User 데이터를 반환하도록 메소드 변경
- 모든 파일 자동 정렬 및 코드 정리

5. 코드 정리 및 주석 추가
추가
- index 쿼리 날린 기록 추가
- sendMessage메소드에 주석 추가

수정
- Entity Class에 Table Name 추가
- 변수 이름을 카멜 케이스로 변경 및 기존 이름은 JsonProperty로 변경
- Transactional Annotation 추가
- 기존의 device와 user의 post method에서 Transactional Annotation에서 Roll Back 시키는 에러 발생. catch에서 error를 제대로 안잡고 return하려고해서 그러는 거 같음. 그래서 findById를 사용하여 isPresent로 id값이 있는지 확인하도록 변경

6. Exception 파일들 error 패키지로 이동 및 PushAlarm 수정
추가
- 알람 보낸 후 callback 메소드에서 log 찍도록 추가
- sendMessage 함수에 Async 어노테이션 추가

수정
- getLogger 클래스를 모든 클래스에서 바로 사용할 수 있도록 변경
- 한 개씩 비동기로 보내던 키워드 알람을 한 번에 모아서 보내도록 수정

remove server url

qwlake reviewed (commented):
코맨트 확인 부탁드립니다

rave189 committed:
JDBCTemplate에서 QueryDSL을 사용하도록 변경
추가
- QueryDSL Dependency 추가

변경
- JDBCTemplate 코드를 QueryDSL 코드로 변경

rave189 committed:
device테이블 분리로 인한 연관관계 매핑

추가
- Keyword, Subcription 엔티티 추가
- Keyword와 Subscription의 Repository 추가
- Device와 Keyword에 1:N관계 설정
- Device와 Subscription에 1:N관계 설정
- Subscription과 Category에 N:1관계 설정
- 기존 GET으로 Device의 ID를 받아서 조회하는 방법이 URL Safe하지 못한 방법이므로 POST 요청으로 조회할 수 있도록 api 추가

수정
- 엔티티 추가에 따른 DeviceDTO 수정
- api 요청 데이터 포맷이 String에서 String 배열로 변경 및 엔티티 추가로 인한 GET, POST, PUT, DELETE 로직 수정

rave189 committed:
공지 관련 컨트롤러 수정

수정
- 데이터의 길이를 나타내는 count가 제대로 들어가지 않던 코드 수정
- 공지사항이 없는 경우도 존재하기 때문에 기존 404 에러를 보내던 잘못된 코드를 삭제

rave189 committed:
공지사항 로직 변경

수정
- 공지사항의 created_at은 공지사항이 서버에 등록된 시각으로 created_at을 내림차순 정렬하면 공지사항이 올라온 날짜 기준으로 정렬이 되지 않음. 따라서 공지사항이 등록된 날짜인 date를 사용하도록 변경

rave189 committed:
알림서버를 위한 모듈 설정

변경
- 알림 서버를 위한 모듈 생성 및 설정 변경
- 모듈 추가에 의한 기존 파일 이동

rave189 committed:
알림 서버 기능 추가 및 공지사항 연관관계 매핑

추가
- 알림을 받을 유저를 선정하여 카프카로 보내는 로직 추가
- 알림 서버에서 카프카의 데이터를 받아 유저에게 보내는 기능 추가

수정
- 기존 api 서버에 있던 FCMConfig를 알림 서버로 이동
- 공지사항의 학과코드와 카테고리 연관관계 설정
- oneToMany가 있는 부분 모두 JsonIgnore 적용(이것 때문에 stackoverflow로 동작을 안함 이유가 뭐지?)
- 공지사항 연관관계 설정에 따른 NoticeRepository, NoticeCustomRepository 수정
- 사용하지 않는 메소드, 라이브러리 및 코드 정리
-

rave189 committed:
JPA Auditing 기능 추가 및 연관관계 FetchType을 Lazy로 변경

추가
- 생성일과 수정일이 들어가는 Entity에 JPA Auditing으로 자동 생성 및 변경되도록 기능 추가

수정
- Entity간 연관관계 매핑 시 FetchType을 Lazy로 바꾸어 연관 객체를 호출할 때 쿼리를 수행하도록 변경

rave189 committed:
Support Id 타입 변경
수정
- 의미없는 Id 값으로 1을 사용하지 않고 version이라는 String 타입으로 변경

rave189 committed:
PushStatus Id 타입 변경

수정
- PushStatus의 의미없는 Id값을 String으로 변경하여 의미있는 Id로 변경
- KafkaConsumer의 ObjectMapper 전역 변수로 변경

rave189 committed:
키워드 알람과 구독 알람을 받도록 설정 했을 시에만 알람을 보내도록 변경

수정
- push 서비스에서 키워드 알람과 구독 알람이 true인 경우에만 kafka에 보내도록 변경
